cpp/src/arrow/compute/exec/expression_test.cc

```diff
@@ -593,8 +594,23 @@ TEST(Expression, BindWithImplicitCasts) {
     ExpectBindsTo(cmp(field_ref("dict_str"), field_ref("str")),
                   cmp(cast(field_ref("dict_str"), utf8()), field_ref("str")));
 
+    // Should prefer the literal
     ExpectBindsTo(cmp(field_ref("dict_i32"), literal(int64_t(4))),
-                  cmp(cast(field_ref("dict_i32"), int64()), literal(int64_t(4))));
+                  cmp(field_ref("dict_i32"), literal(int32_t(4))));
+    ExpectBindsTo(cmp(field_ref("dict_i32"), literal(int64_t(4))),
```

Comment:
Please add a case which demonstrates the behavior in the presence of unsigned integers. `SmallestTypeFor` preserves signedness information, which can produce some odd edge cases which we should be explicit about. For example:

```suggestion
    ExpectBindsTo(cmp(field_ref("i8"), literal(uint8_t(4))),
                  cmp(cast(field_ref("i8"), int16()), literal(int16_t(4))));
    ExpectBindsTo(cmp(field_ref("dict_i32"), literal(int64_t(4))),
```

Reply:
Done.